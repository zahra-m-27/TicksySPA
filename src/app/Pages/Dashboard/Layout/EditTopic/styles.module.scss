.container {
  flex: 1;
  padding: 25px 50px;
}
