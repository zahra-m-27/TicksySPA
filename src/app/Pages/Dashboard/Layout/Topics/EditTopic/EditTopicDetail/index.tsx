import {message, Spin} from 'antd';
import API from '../../../../../../API';
import Assets from '../../../../../../Assets';
import styles from './styles.module.scss';
import {useEffect, useRef, useState} from 'react';
import SEInput from '../../../../../../Components/SEInput';
import {useHistory, useParams} from 'react-router-dom';
import EditTopicCard from '../../../../../../Components/EditTopicCard';

export default function EditTopicDetail() {
  const history = useHistory();
  const params = useParams<{topicId: string}>();
  const [Avatar, setAvatar] = useState<File>();
  const [Title, setTitle] = useState('');
  const [AvatarUrl, setAvatarUrl] = useState<any>();
  const [Description, setDescription] = useState('');
  const [SaveLoading, setSaveLoading] = useState(false);
  const descriptionRef = useRef<HTMLTextAreaElement>(null);
  const [TitleHasError, setTitleHasError] = useState(false);
  const [Loading, setLoading] = useState(!process.env.JEST_WORKER_ID);
  const [DescriptionHasError, setDescriptionHasError] = useState(false);

  useEffect(() => {
    setLoading(!process.env.JEST_WORKER_ID);
    API.Topics.GetTopic({
      topicId: parseInt(params.topicId),
    })
      .then((response) => {
        setTitle(response.title);
        setAvatarUrl(response.avatar);
        setDescription(response.description);
      })
      .catch(() => message.error('اشکالی در دریافت اطلاعات تایپک رخ داده.'))
      .finally(() => setLoading(false));
  }, []);

  const onEditTopic = () => {
    if (!Title) {
      setTitleHasError(true);
    }
    if (!Description) {
      setDescriptionHasError(true);
    }
    if (!Avatar && !AvatarUrl) {
      message.error('برای تاپیک خود یک آواتار انتخاب کنید');
    }
    if (!Title || !Description || !Avatar) {
      return;
    }
    setTitleHasError(false);
    setDescriptionHasError(false);

    setSaveLoading(true);
    API.Topics.UpdateTopic({
      title: Title,
      avatar: Avatar,
      description: Description,
      topicId: parseInt(params.topicId),
    })
      .then(() => {
        message.success('تاپیک با بروز شد.');
        history.push('/dashboard/topics');
      })
      .catch((error) => {
        if (error.status === 403) {
          message.error('برای این عملیات نیاز هست تا احراز هویت کرده باشید');
        } else {
          message.error('انجام این عملیات ممکن نیست');
        }
      })
      .finally(() => setSaveLoading(false));
  };

  const onTitleEnter = () => {
    if (descriptionRef && descriptionRef.current) {
      descriptionRef.current.focus();
    }
  };

  if (Loading) {
    return (
      <div className={styles.container}>
        <Spin size="large" />
      </div>
    );
  }

  return (
    <div className={styles.container}>
      <EditTopicCard
        title="ویرایش تاپيک"
        icon={Assets.SVGs.Topic}
        className={styles.detail_card}
        buttons={[
          {
            label: 'ثبت',
            testId: 'submit',
            onClick: onEditTopic,
            loading: SaveLoading,
            className: styles.enter_button,
          },
        ]}>
        <div className={styles.upload}>
          <label htmlFor="picture">
            <img
              alt="avatar"
              className={styles.picture}
              src={AvatarUrl ?? Assets.Images.GoogleImage2}
            />
          </label>
          <input
            type="file"
            id="picture"
            data-testid="avatar"
            className={styles.upload_image}
            onChange={(e) => {
              if (e.target.files) {
                setAvatar(e.target.files[0]);
                const fr = new FileReader();
                fr.onload = function () {
                  setAvatarUrl(fr.result);
                };
                fr.readAsDataURL(e.target.files[0]);
              }
            }}
          />
        </div>
        <div className={styles.inputs_container}>
          <div className={styles.middle}>
            <label>عنوان</label>
            <SEInput
              content={Title}
              data-testid="title"
              onEnter={onTitleEnter}
              onChangeText={setTitle}
              hasError={TitleHasError}
              innerContainerClassName={styles.input_container}
            />
          </div>
          <div className={styles.below}>
            <label>توضيح</label>
            <SEInput
              minLines={5}
              ref={descriptionRef}
              content={Description}
              data-testid="description"
              onChangeText={setDescription}
              hasError={DescriptionHasError}
              inputClassName={styles.description_input}
              innerContainerClassName={styles.input_container}
            />
          </div>
        </div>
      </EditTopicCard>
    </div>
  );
}
