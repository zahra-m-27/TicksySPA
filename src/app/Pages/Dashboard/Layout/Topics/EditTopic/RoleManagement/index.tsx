import {message, Spin} from 'antd';
import API from '../../../../../../API';
import styles from './styles.module.scss';
import {useParams} from 'react-router-dom';
import Assets from '../../../../../../Assets';
import React, {MutableRefObject, useEffect, useRef, useState} from 'react';
import TButton from '../../../../../../Components/TButton';
import showDialog from '../../../../../../Components/TDialog';
import ManageTopicRoleDialog from '../../../../../../Dialogs/ManageTopicRole';
import TPagination from '../../../../../../Components/TPagination';
import TopicAdminListItemDto from '../../../../../../API/DTOs/TopicAdminListItemDto';

interface Props {
  onUpdate: MutableRefObject<(() => void) | undefined>;
}

export default function RoleManagement({onUpdate}: Props) {
  const dismissDialog = useRef<() => void>();
  const params = useParams<{topicId: string}>();

  const [Total, setTotal] = useState(1);
  const [PageSize, setPageSize] = useState(5);
  const [PageNumber, setPageNumber] = useState(0);
  const [Roles, setRoles] = useState<TopicAdminListItemDto[]>();

  const getData = () => {
    API.Topics.GetTopicRoles({
      limit: PageSize,
      offset: PageNumber * PageSize,
      topicId: parseInt(params.topicId),
    })
      .then((response) => {
        setTotal(response.count);
        setRoles(
          response.results.concat(
            new Array(
              (PageSize - (response.results.length % PageSize)) % PageSize
            ).fill({})
          )
        );
      })
      .catch(() => {
        message.error('خطایی در گرفتن نقش های تاپیک رخ داده.');
      });
  };

  onUpdate.current = getData;
  useEffect(getData, [PageNumber]);

  let content = <Spin size="large" />;

  if (Roles)
    content = (
      <div className={styles.not_data}>
        <Assets.SVGs.SadSvg />
        <span className={styles.not_data_text}>دسته بندی ای یافت نشد</span>

        <TButton
          backgroundColor="#1e77c9"
          label="نقش جدید ایجاد کن"
          className={styles.create_no_data}
          labelClassName={styles.create_no_data_label}
          onClick={() =>
            (dismissDialog.current = showDialog({
              content: (
                <ManageTopicRoleDialog
                  onUpdate={getData}
                  onDismissRef={dismissDialog}
                  topicId={parseInt(params.topicId)}
                />
              ),
            }))
          }
        />
      </div>
    );

  if (Roles?.length)
    content = (
      <>
        <table className={styles.table}>
          <thead>
            <tr>
              <th>ردیف</th>
              <th>عنوان نقش</th>
              <th>تعداد مسئولان</th>
              <th>عملیات</th>
            </tr>
          </thead>
          <tbody>
            {Roles.map((role, index) => {
              if (!role.id) return <tr />;
              return (
                <tr key={index}>
                  <td>{role.id}</td>
                  <td>{role.title}</td>
                  <td>{role.users_detail.length}</td>
                  <td>
                    <img
                      alt="edit"
                      src={Assets.Images.EditUsers}
                      className={styles.edit_icon}
                      onClick={() =>
                        (dismissDialog.current = showDialog({
                          content: (
                            <ManageTopicRoleDialog
                              roleId={role.id}
                              onUpdate={getData}
                              title={role.title}
                              onDismissRef={dismissDialog}
                              topicId={parseInt(params.topicId)}
                              users={role.users_detail.map((u) => u.id)}
                            />
                          ),
                        }))
                      }
                    />
                    <img
                      alt="delete"
                      onClick={() => {
                        API.Topics.DeleteRole({
                          roleId: role.id,
                          topicId: parseInt(params.topicId),
                        })
                          .then(() => {
                            message.success(`نقش ${role.title} حذف شد.`);
                          })
                          .catch(() => {
                            message.error('خطایی در حذف نقش ها رخ داده.');
                          })
                          .finally(getData);
                      }}
                      src={Assets.Images.Delete}
                      className={styles.delete_icon}
                    />
                  </td>
                </tr>
              );
            })}
          </tbody>
        </table>
        <TPagination
          total={Total}
          pageSize={PageSize}
          pageNumber={PageNumber + 1}
          onChange={(value) => setPageNumber(value - 1)}
        />
      </>
    );

  return <div className={styles.container}>{content}</div>;
}
