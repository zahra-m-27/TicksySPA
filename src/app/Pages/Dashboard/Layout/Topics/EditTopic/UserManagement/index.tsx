import {message, Spin} from 'antd';
import Assets from '../../../../../../Assets';
import styles from './styles.module.scss';
import showDialog from '../../../../../../Components/TDialog';
import ManageTopicUserDialog from '../../../../../../Dialogs/ManageTopicUser';
import React, {MutableRefObject, useEffect, useRef, useState} from 'react';
import {useParams} from 'react-router-dom';
import API from '../../../../../../API';
import {StarTwoTone} from '@ant-design/icons';
import TButton from '../../../../../../Components/TButton';
import TPagination from '../../../../../../Components/TPagination';
import TopicUsersListSerializerDto from '../../../../../../API/DTOs/TopicUsersListSerializerDto';

interface Props {
  onUpdate: MutableRefObject<(() => void) | undefined>;
}

export default function UserManagement({onUpdate}: Props) {
  const dismissDialog = useRef<() => void>();
  const params = useParams<{topicId: string}>();

  const [Total, setTotal] = useState(1);
  const [PageSize, setPageSize] = useState(5);
  const [PageNumber, setPageNumber] = useState(0);
  const [Users, setUsers] = useState<TopicUsersListSerializerDto[]>();

  const getData = () => {
    API.Topics.GetTopicUsers({
      limit: PageSize,
      offset: PageNumber * PageSize,
      topicId: parseInt(params.topicId),
    })
      .then((response) => {
        setTotal(response.count);
        setUsers(
          response.results.concat(
            new Array(
              (PageSize - (response.results.length % PageSize)) % PageSize
            ).fill({})
          )
        );
      })
      .catch(() => message.error('خطایی در گرفتن نقش های تاپیک رخ داده.'));
  };

  onUpdate.current = getData;
  useEffect(getData, [PageNumber]);

  let content = <Spin size="large" />;

  if (Users)
    content = (
      <div className={styles.not_data}>
        <Assets.SVGs.SadSvg />
        <span className={styles.not_data_text}>دسته بندی ای یافت نشد</span>

        <TButton
          backgroundColor="#1e77c9"
          label="کاربر جدید ایجاد کن"
          className={styles.create_no_data}
          labelClassName={styles.create_no_data_label}
          onClick={() =>
            (dismissDialog.current = showDialog({
              content: (
                <ManageTopicUserDialog
                  onUpdate={getData}
                  onDismissRef={dismissDialog}
                  topicId={parseInt(params.topicId)}
                />
              ),
            }))
          }
        />
      </div>
    );

  if (Users?.length)
    content = (
      <>
        <table className={styles.table}>
          <thead>
            <tr>
              <th>ردیف</th>
              <th>نام نام خانوادگی</th>
              <th>عنوان نقش</th>
              <th>ایمیل</th>
              <th>عملیات</th>
            </tr>
          </thead>
          <tbody>
            {Users.map((user, index) => {
              if (!user.id) return <tr />;
              return (
                <tr key={index}>
                  <td>{user.id}</td>
                  <td>
                    <div className={styles.user_name}>
                      <img
                        src={user.avatar ?? Assets.SVGs.MaleUserSVG}
                        alt="avatar"
                      />
                      <span>
                        {user.first_name} {user.last_name}
                      </span>
                      {user.is_creator && <StarTwoTone />}
                    </div>
                  </td>
                  <td>
                    {user.admin_set
                      .map((a) => a.title)
                      .concat(
                        user.is_creator && user.admin_set.length === 0
                          ? ['مدیر اصلی تاپیک']
                          : []
                      )
                      .join(', ')}
                  </td>
                  <td>{user.email}</td>
                  <td>
                    <img
                      alt="edit"
                      data-testid="edit_user"
                      src={Assets.Images.EditUsers}
                      className={styles.edit_icon}
                      onClick={() =>
                        (dismissDialog.current = showDialog({
                          content: (
                            <ManageTopicUserDialog
                              user={user}
                              onUpdate={getData}
                              onDismissRef={dismissDialog}
                              topicId={parseInt(params.topicId)}
                            />
                          ),
                        }))
                      }
                    />
                  </td>
                </tr>
              );
            })}
          </tbody>
        </table>
        <TPagination
          total={Total}
          pageSize={PageSize}
          pageNumber={PageNumber + 1}
          onChange={(value) => setPageNumber(value - 1)}
        />
      </>
    );
  return <div className={styles.container}>{content}</div>;
}
