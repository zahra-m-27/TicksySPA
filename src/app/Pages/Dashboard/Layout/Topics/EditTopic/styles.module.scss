.container {
  flex: 1;
  padding: 25px 50px;
}

.add_button {
  padding: 5px 10px;
  margin-bottom: 5px;

  span {
    font-size: clamp(12px, 1vw, 16px);
  }
}
