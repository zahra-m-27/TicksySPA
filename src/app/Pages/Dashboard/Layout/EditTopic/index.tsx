import styles from './styles.module.scss';
import {Tabs} from 'antd';
import {useState} from 'react';
import EditTopicDetail from './EditTopicDetail';
import UserManagement from '../UserManagement';
import RoleManagement from '../RoleManagement';

export default function EditTopic() {
  const [Tab, setTab] = useState('detail');

  return (
    <div className={styles.container}>
      <Tabs accessKey={Tab} onChange={(k) => setTab(k)}>
        <Tabs.TabPane tab="مشخصات" key="detail">
          <EditTopicDetail />
        </Tabs.TabPane>
        <Tabs.TabPane tab="کاربران" key="users">
          <UserManagement />
        </Tabs.TabPane>
        <Tabs.TabPane tab="نقش ها" key="roles">
          <RoleManagement />
        </Tabs.TabPane>
      </Tabs>
    </div>
  );
}
