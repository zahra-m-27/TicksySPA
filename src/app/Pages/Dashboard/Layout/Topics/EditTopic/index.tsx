import styles from './styles.module.scss';
import {Button, Tabs} from 'antd';
import React, {useRef, useState} from 'react';
import EditTopicDetail from './EditTopicDetail';
import UserManagement from './UserManagement';
import RoleManagement from './RoleManagement';
import showDialog from '../../../../../Components/TDialog';
import ManageTopicRoleDialog from '../../../../../Dialogs/ManageTopicRole';
import TButton from '../../../../../Components/TButton';
import ManageTopicUserDialog from '../../../../../Dialogs/ManageTopicUser';
import {useParams} from 'react-router-dom';

export default function EditTopic() {
  const updateRoles = useRef<() => void>();
  const updateUsers = useRef<() => void>();
  const dismissDialog = useRef<() => void>();
  const params = useParams<{topicId: string}>();
  const [Tab, setTab] = useState('detail');

  let addTitle = '';
  if (Tab === 'roles') {
    addTitle = 'افزودن نقش';
  } else if (Tab === 'users') {
    addTitle = 'افزودن کاربر';
  }

  return (
    <div className={styles.container}>
      <Tabs
        accessKey={Tab}
        onChange={setTab}
        tabBarExtraContent={
          addTitle && {
            right: (
              <TButton
                label={addTitle}
                className={styles.add_button}
                onClick={() => {
                  if (Tab === 'roles')
                    dismissDialog.current = showDialog({
                      content: (
                        <ManageTopicRoleDialog
                          onDismissRef={dismissDialog}
                          onUpdate={updateRoles.current}
                          topicId={parseInt(params.topicId)}
                        />
                      ),
                    });
                  else
                    dismissDialog.current = showDialog({
                      content: (
                        <ManageTopicUserDialog
                          onDismissRef={dismissDialog}
                          onUpdate={updateUsers.current}
                          topicId={parseInt(params.topicId)}
                        />
                      ),
                    });
                }}
              />
            ),
          }
        }>
        <Tabs.TabPane tab="مشخصات" key="detail">
          <EditTopicDetail />
        </Tabs.TabPane>
        <Tabs.TabPane tab="کاربران" key="users">
          <UserManagement onUpdate={updateUsers} />
        </Tabs.TabPane>
        <Tabs.TabPane tab="نقش ها" key="roles">
          <RoleManagement onUpdate={updateRoles} />
        </Tabs.TabPane>
      </Tabs>
    </div>
  );
}
