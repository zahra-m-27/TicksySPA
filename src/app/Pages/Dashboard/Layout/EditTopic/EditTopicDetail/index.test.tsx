import ReactDOM from 'react-dom';
import React from 'react';
import {render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';
import EditTopic from './index';

const sample = (
  <MemoryRouter>
    <EditTopic />
  </MemoryRouter>
);

test('render in ReactDOM', () => {
  const div = document.createElement('div');
  ReactDOM.render(sample, div);
});

it('render member header', () => {
  const {getByText} = render(sample);
  getByText(/اعضا/i);
});

it('render detail header', () => {
  const {getByText} = render(sample);
  getByText(/ویرایش تاپيک/i);
});

it('render add member', () => {
  const {getByText} = render(sample);
  getByText(/افزودن عضو/i);
});

it('render subject', () => {
  const {getByText} = render(sample);
  getByText(/عنوان/i);
});

it('render username', () => {
  const {getByText} = render(sample);
  getByText(/شناسه/i);
});

it('render description', () => {
  const {getByText} = render(sample);
  getByText(/توضيح/i);
});

it('render submit', () => {
  const {getByText} = render(sample);
  getByText(/ثبت/i);
});

it('snapshot', () => {
  const snapshot = renderer.create(sample).toJSON();
  expect(snapshot).toMatchSnapshot();
});
