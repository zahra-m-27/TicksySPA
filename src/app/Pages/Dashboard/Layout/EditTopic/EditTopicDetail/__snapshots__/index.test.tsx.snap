// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="container"
>
  <div
    className="card member_card"
  >
    <div
      className="header"
    >
      اعضا
      <img
        src="people.svg"
      />
    </div>
    <div
      className="content member_content"
    >
      <div
        className="member_container add_button"
        onClick={[Function]}
      >
        <img
          alt="add member"
          src="plus.svg"
        />
        <p>
          افزودن عضو
        </p>
      </div>
    </div>
  </div>
  <div
    className="card detail_card"
  >
    <div
      className="header"
    >
      ویرایش تاپيک
      <img
        src="topic.svg"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="upload"
      >
        <label
          htmlFor="picture"
        >
          <img
            alt="avatar"
            className="picture"
            src="google_images_2.png"
          />
        </label>
        <input
          className="upload_image"
          id="picture"
          onChange={[Function]}
          type="file"
        />
      </div>
      <div
        className="inputs_container"
      >
        <div
          className="middle"
        >
          <label>
            :عنوان
          </label>
          <div
            className="input_container"
          >
            <div
              className="inner_container input_container"
            >
              <div
                className="content"
              >
                <input
                  className="input"
                  dir="auto"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <label>
            :شناسه
          </label>
          <div
            className="input_container"
          >
            <div
              className="inner_container input_container"
            >
              <div
                className="content"
              >
                <input
                  className="input"
                  dir="auto"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="below"
        >
          <label>
            :توضيح
          </label>
          <div
            className="input_container"
          >
            <div
              className="inner_container input_container"
            >
              <div
                className="content"
              >
                <textarea
                  className="input description_input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder=""
                  rows={5}
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      className="ant-btn ant-btn-primary enter_button"
      onClick={[Function]}
      type="button"
    >
      <span>
        ثبت
      </span>
    </button>
  </div>
</div>
`;
